.github/workflows/template.yml

```
@@ -0,0 +1,96 @@
+name: Template
+on:
+  push:
+    branches: gh-pages
+  pull_request: []
+jobs:
+  check-template:
+    name: Test lesson template
+    if: github.repository == 'carpentries/styles'
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        lesson: [swcarpentry/shell-novice, datacarpentry/r-intro-geospatial, librarycarpentry/lc-git]
+        os: [ubuntu-latest, macos-latest, windows-latest]
+    defaults:
+      run:
+        shell: bash # forces 'Git for Windows' on Windows
+    steps:
+      - name: Set up Ruby
+        uses: actions/setup-ruby@main
+        with:
+          ruby-version: '2.7.1'
+
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.x'
+
+      - name: Install GitHub Pages, Bundler, and kramdown gems
+        run: |
+          gem install github-pages bundler kramdown
+
+      - name: Install Python modules
+        run: |
+          if [[ $RUNNER_OS == macOS || $RUNNER_OS == Linux ]]; then
+            python3 -m pip install --upgrade pip setuptools wheel pyyaml==5.3.1 requests
+          elif [[ $RUNNER_OS == Windows ]]; then
+            python -m pip install --upgrade pip setuptools wheel pyyaml==5.3.1 requests
+          fi
+
+      - name: Checkout the ${{ matrix.lesson }} lesson
+        uses: actions/checkout@master
+        with:
+          repository: ${{ matrix.lesson }}
+          path: lesson
+          fetch-depth: 0
+
+      - name: Look for R-markdown files
+        id: check-rmd
+        working-directory: lesson
+        run: |
+          echo "::set-output name=count::$(shopt -s nullglob; files=($(find . -iname '*.Rmd')); echo ${#files[@]})"
+
+      - name: Set up R
+        if: steps.check-rmd.outputs.count != 0
+        uses: r-lib/actions/setup-r@master
+        with:
+          r-version: 'release'
+
+      - name: Install R packages
+        if: steps.check-rmd.outputs.count != 0
+        run: |
+          install.packages('remotes', 'rprojroot', 'renv')
+        shell: Rscript {0}
+
+      - name: Cache R packages
+        if: runner.os != 'Windows' && steps.check-rmd.outputs.count != 0
+        uses: actions/cache@v1
+        with:
+          path: ${{ env.R_LIBS_USER }}
+          key: ${{ runner.os }}-${{ hashFiles('.github/R-version') }}-1-${{ hashFiles('.github/depends.Rds') }}
+          restore-keys: ${{ runner.os }}-${{ hashFiles('.github/R-version') }}-1-
```

@fmichonneau, I believe we have to swap the order of "Cache R packages" and "Install R packages" and check for cache hits (`steps.cache-r.outputs.cache-hit != 'true'`):
```suggestion
      - name: Cache R packages
        id: cache-r
        if: runner.os != 'Windows' && steps.check-rmd.outputs.count != 0
        uses: actions/cache@v1
        with:
          path: ${{ env.R_LIBS_USER }}
          key: ${{ runner.os }}-${{ hashFiles('.github/R-version') }}-1-${{ hashFiles('.github/depends.Rds') }}
          restore-keys: ${{ runner.os }}-${{ hashFiles('.github/R-version') }}-1-
          
      - name: Install R packages
        if: steps.cache-r.outputs.cache-hit != 'true' && steps.check-rmd.outputs.count != 0
        run: |
          install.packages('remotes', 'rprojroot', 'renv')
        shell: Rscript {0}
```

Please approve/comment.

---

these are the packages that are needed for all the R-based lessons. The idea is that these packages would always be installed fresh but the packages that are lesson-specific get cached.

---

Currently `Cache R packages` step is not used -- the idea is to skip package installation step when `steps.cache-r.outputs.cache-hit == 'true`.

Don't we also need 'knitr', 'stringr', 'checkpoint', 'ggplot2'?

---

not anymore, we now use a combination of `renv` + `rprojroot` which takes care of identifying all the dependencies in our lessons that are then installed as part of the build process.

---

> not anymore, we now use a combination of `renv` + `rprojroot` which takes care of identifying all the dependencies in our lessons that are then installed as part of the build process.

Great!

So, right now the caching step is used to cache downloads only and that's exactly what you want. I'm 👍 with that.

---

wait, I think current order has to be switched, though -- "Install R packages" step won't see any cache because it's created only after it is executed. Is there a way to download R packages before installing them?

---

we can have `cache-restore` and `cache-save` steps before and after we install R packages.

---

Resolved in https://github.com/maxim-belkin/styles/pull/3